Document Metadata structure (#861)
PixarAnimationStudios/OpenTimelineIO: docs/tutorials/otio-file-format-specification.md

## Version

> This DRAFT describes the OpenTimelineIO JSON File Format as of OTIO Beta 13.

Reviewer: Is there a way to embed this into the automatically generated document?  Otherwise we should add to our release process updating this document.

Author: Given that we didn't actually update the contents of this since Alpha 5, I think it was correct. We should add a manual process to any schema enhancements to include documentation.

Reviewer: Sorry, to include updating the documentation?  Or to include documentation in general?  

Either way, we should keep a list of which pages to check then.  We can add a todo to our TSC documentation pass list.

Author: I think both. Specifically any PR that changes schema should update this document (in addition to the autogenerated one) - and when we do a release, we should have a reminder "check that the file format spec documentation reflects any schema changes" in case we missed anything important.

Reviewer: Makes sense.  Thanks!

The top level object in an OTIO file can be any OTIO data type, but is typically a Timeline. This means that most use cases will assume that the top level

> object is a Timeline, but in specific workflows, you can read and write otio files that contain

Reviewer: ```suggestion
object is a Timeline, but in specific workflows, otio files can be read or written that contain
```

> just a Clip, Track, RationalTime, or any other OTIO data type. Due to the nature of JSON, you could also read/write an array of objects, but we

Reviewer: ```suggestion
just a Clip, Track, RationalTime, or any other OTIO data type. Due to the nature of JSON, arrays of objects can also be read/written, but we
```

> recommend that you use the OTIO SerializableCollection data type in this case so that you can attach metadata to the container itself. Code that reads an OTIO

Reviewer: ```suggestion
recommend using the OTIO SerializableCollection data type in this case so that metadata can be attached to the container itself. Code that reads an OTIO
```

or dictionary. If the value is a dictionary, then it will often be an OTIO data type. In some cases (specifically metadata) it can be a regular JSON dictionary.

OTIO JSON files are typically formatted with indentation to make them easier to read. This makes the files slightly larger, but dramatically improves human
readability which makes debugging much easier. Furthermore, the OTIO library will write the keys of each object in a predictable order to help with change tracking, comparisons, etc.

> Since human readablility and ease of use are explicit goals of the OpenTimelineIO project, we recommend that you do not minify OTIO JSON unless absolutely necessary. If file size is really important, you should probably gzip them instead.

Reviewer: ```suggestion
Since human readablility and ease of use are explicit goals of the OpenTimelineIO project, it is recommended that OTIO JSON not be minified unless absolutely necessary. If a minimum file size is desired, the recommendation is to use gzip rather than minifying.
```

## Metadata

> Timeline, Stack, Track, Clip, MediaReferece, and most other OTIO objects all have a `metadata` property.

Reviewer 2: It may be worth explicitly calling out here that the `metadata` property is an arbitrarily deep JSON dictionary like structure with no enforced schema.

> This is intended to be a place for you to put information that does not fit into the standard properties.

Reviewer: ```suggestion
This is intended to be a place to put information that does not fit into the schema defined properties.
```

The core of OTIO doesn't do anything with metadata, it only carries it along so that adapters, scripts,

> applications, or other workflows can use that metadata however needed. For example, you will find that

Reviewer: ```suggestion
applications, or other workflows can use that metadata however needed. For example, 
```

several of the adapters shipped with OTIO use metadata to store information that doesn't (yet) fit into
the core OTIO schema.

Due to the fact that many different workflows can and will use metadata, it is important to group your

> metadata inside a namespace so that all of those can coexist without stomping on each other. In the example below, you can see that there is metadata on the Timeline and on several Clips for both a hypothetical `my_playback_tool` and `my_production_tracking_system` that could coexist with anything else you add under a similar namespace.

Reviewer: ```suggestion
metadata inside a namespace so that all of those can coexist without encountering a name collision. In the example below, there is metadata on the Timeline and on several Clips for both a hypothetical `my_playback_tool` and `my_production_tracking_system` that could coexist with anything else added under a similar namespace.
```

> Metadata can also be useful when prototyping new OTIO schemas. An existing object can be extended with metadata which can later be upgraded into a new schema version, or a schemadef plugin.

Reviewer 2: Perhaps "migrated into top-level attributes on a new schema version" would be clearer than "upgraded" to convey that the data should move out of the `metadata` dictionary and on to "top-level" properties.
